Properly infer intervals of pcolormesh when plotting on logscale

santisoler opened the pull request:
When inferring the interval breaks, the `coord` and first linearized if
`scale="log"` and then the intervals are brought back to the original scale
before being returned to `pcolormesh`. Raise an error if at least one of the
elements of the coodinates has a non-positive element when using a logscale.

- [x] Closes #5333
- [x] Tests added
- [x] Passes `pre-commit run --all-files`
- [x] User visible changes (including notable bug fixes) are documented in `whats-new.rst`
- [x] New functions/methods are listed in `api.rst`

shoyer commented:
this is great, thank you!

could you please look into adding a unit-test, to ensure that this stays fixed?

see http://xarray.pydata.org/en/stable/contributing.html#writing-tests

santisoler commented:
Thanks @shoyer!

> could you please look into adding a unit-test, to ensure that this stays fixed?

Yes! I got a little bit busier after I opened this PR and couldn't find the time to do so. Let me find a gap to add some tests. I'll ping you as soon as the PR is ready for a review. Thanks!

santisoler commented:
@shoyer Now I think this PR is ready to be reviewed. I added one test function for the `_infer_interval_breaks` private function with the optional `scale="log"` argument. I've also added a new test class for checking if calling `pcolormesh` with both `xscale="log"` and `yscale="log"` works as expected. I added the bug fix to `whats-new.rst` and run `pre-commit run --all-files` locally with success. I don't think we need to change anything in `api.rst`.

Let me know if you think I should change or add anything else, and feel free to make any suggestion on the review.

santisoler commented:
@shoyer I've just fixed some errors on the `TestAxesKwargs` tests that were using logscales without passing any coordinates.
I've also added an error raising if the passed coordinates have at least one non-positive element.
Hope CIs pass now. Let me know if you feel that something else should be changed.

shoyer commented on doc/whats-new.rst:
```diff
@@ -45,6 +45,9 @@ Deprecations
 Bug fixes
 ~~~~~~~~~
 
+- Plotting a pcolormesh with ``xscale="log"`` and/or ``yscale="log"`` works as
+  expected after improving the way the interval breaks are generated (:issue:`5333`).
+  By `Santiago Soler <https://github.com/santisoler>`_
```
When you get the chance, please merge in the latest changes from git master and move this to the section for 0.18.3.

shoyer reviewed (approved):
This looks great, thank you!

max-sixty commented on xarray/tests/test_plot.py:
```diff
@@ -2638,6 +2717,16 @@ def test_ncaxis_notinstalled_line_plot(self):
 ]
 
+# Define a few arrays for testing scale="log"
+test_da_list_log = [
```
@santisoler please could we make this into a fixture? If it's in the base of the module, then it'll be run on every import of this file, which will slow down collecting tests (even those that don't use it)

santisoler replied:
I've just transformed the `test_da_list_log` as a fixture method in `TestAxesKwargs` with the number of dimensions of the returned DataArray as a parameter of the fixture. I took the liberty to do the same with the preexisting `test_da_list `.

shoyer reviewed (approved):
LGTM

@max-sixty any remaining concerns with the test fixtures?

dcherian commented:
I think we can address any concerns about test fixtures in a future PR.

Thanks @santisoler. I see this is your first contribution. welcome to xarray!

santisoler commented:
Thanks @dcherian! I really liked the PR process in xarray, having 3 reviewers here was awesome and a crucial part for building quality code. Thanks for your hard work and comments! Hope this to be the first PR or many more!
